File: cli/npm/src/commands/explain.ts
import { Command } from 'commander';
import chalk from 'chalk';
import * as fs from 'fs-extra';
import * as path from 'path';
import { PrompdParser } from '../lib/parser';
import { RegistryClient } from '../lib/registry';

export function createExplainCommand(): Command {
  const explainCommand = new Command('explain');

  explainCommand
    .description('Detailed information about files and packages')
    .argument('<target>', '.prmd file, .pdpkg package, or registry package (@namespace/name)')
    .option('-d, --detailed', 'Show detailed information')
    .option('-s, --sections', 'Show section content previews (for .prmd files)')
    .option('-h, --history', 'Show git version history (for .prmd files)')
    .option('-r, --registry <name>', 'Specify registry for package lookup')
    .option('-v, --verbose', 'Enable verbose logging')
    .action(async (target: string, options) => {
      try {
        // Determine target type
        if (target.endsWith('.prmd')) {
          await explainPrompdFile(target, options);
        } else if (target.endsWith('.pdpkg')) {
          await explainPackageFile(target, options);
        } else if (target.startsWith('@') && target.includes('/')) {
          await explainRegistryPackage(target, options);
        } else {
          console.error(chalk.red('Invalid target. Must be a .prmd file, .pdpkg package, or registry package (@namespace/name)'));
          process.exit(1);
        }
      } catch (error) {
        console.error(chalk.red(`Explain failed: ${error}`));
        if (options.verbose) {
          console.error(error);
        }
        process.exit(1);
      }
    });

  return explainCommand;
}

async function explainPrompdFile(filePath: string, options: any) {
  if (!await fs.pathExists(filePath)) {
    throw new Error(`File not found: ${filePath}`);
  }

  const parser = new PrompdParser();
  const prompd = await parser.parseFile(filePath);
  const stats = await fs.stat(filePath);

  console.log(chalk.bold(`\n📄 ${path.basename(filePath)}`));
  console.log(chalk.dim(`   ${path.resolve(filePath)}`));
  console.log();

  // Basic metadata
  console.log(chalk.bold('Metadata:'));
  console.log(`  ID:          ${chalk.cyan(prompd.metadata.id)}`);
  if (prompd.metadata.name) {
    console.log(`  Name:        ${prompd.metadata.name}`);
  }
  if (prompd.metadata.version) {
    console.log(`  Version:     ${prompd.metadata.version}`);
  }
  if (prompd.metadata.description) {
    console.log(`  Description: ${prompd.metadata.description}`);
  }
  console.log();

  // Parameters
  const params = prompd.metadata.parameters || [];
  if (params.length > 0) {
    console.log(chalk.bold('Parameters:'));
    params.forEach(param => {
      const required = param.required ? chalk.red('required') : chalk.dim('optional');
      const defaultVal = param.default !== undefined ? chalk.dim(` (default: ${JSON.stringify(param.default)})`) : '';
      console.log(`  • ${chalk.cyan(param.name)} (${param.type}) ${required}${defaultVal}`);
      if (param.description) {
        console.log(`    ${chalk.dim(param.description)}`);
      }
    });
    console.log();
  }

  // Sections
  if (options.sections && Object.keys(prompd.sections).length > 0) {
    console.log(chalk.bold('Sections:'));
    Object.keys(prompd.sections).forEach(section => {
      console.log(`  • ${chalk.cyan(section)}`);
    });
    console.log();
  }

  // File stats
  console.log(chalk.bold('File Info:'));
  console.log(`  Size:     ${chalk.cyan(formatBytes(stats.size))}`);
  console.log(`  Modified: ${chalk.dim(stats.mtime.toLocaleString())}`);
  console.log();

  // Git history (if requested)
  if (options.history) {
    try {
      const { spawn } = await import('child_process');
      const gitLog = await new Promise<string>((resolve, reject) => {
        const child = spawn('git', ['log', '--oneline', '-5', filePath]);
        let output = '';
        child.stdout.on('data', (data) => output += data.toString());
        child.on('close', (code) => {
          if (code === 0) resolve(output);
          else reject(new Error('Git log failed'));
        });
      });

      if (gitLog.trim()) {
        console.log(chalk.bold('Recent Git History:'));
        gitLog.trim().split('\n').forEach(line => {
          console.log(`  ${chalk.dim(line)}`);
        });
        console.log();
      }
    } catch (error) {
      console.log(chalk.yellow('Git history not available'));
      console.log();
    }
  }
}

async function explainPackageFile(filePath: string, options: any) {
  if (!await fs.pathExists(filePath)) {
    throw new Error(`Package not found: ${filePath}`);
  }

  const stats = await fs.stat(filePath);

  console.log(chalk.bold(`\n📦 ${path.basename(filePath)}`));
  console.log(chalk.dim(`   ${path.resolve(filePath)}`));
  console.log();

  console.log(chalk.bold('Package Info:'));
  console.log(`  Size: ${chalk.cyan(formatBytes(stats.size))}`);
  console.log(`  Modified: ${chalk.dim(stats.mtime.toLocaleString())}`);
  console.log();

  if (options.detailed) {
    console.log(chalk.yellow('Use "prompd package validate" for detailed package inspection'));
    console.log();
  }
}

async function explainRegistryPackage(packageName: string, options: any) {
  const client = new RegistryClient(options.registry);

  console.log(chalk.dim(`Fetching package info for ${packageName}...`));
  console.log();

  const info = await client.getPackageInfo(packageName);
  const versions = await client.getPackageVersions(packageName);

  console.log(chalk.bold(`\n📦 ${packageName}`));
  console.log();

  if (info.description) {
    console.log(info.description);
    console.log();
  }

  console.log(chalk.bold('Package Info:'));
  console.log(`  Name:    ${chalk.cyan(info.name)}`);
  if (info.version) {
    console.log(`  Version: ${chalk.cyan(info.version)}`);
  }
  console.log();

  if (versions && versions.length > 0) {
    console.log(chalk.bold('Available Versions:'));
    const sortedVersions = versions.sort((a: string, b: string) => {
      return b.localeCompare(a, undefined, { numeric: true, sensitivity: 'base' });
    });
    const displayVersions = options.detailed ? sortedVersions : sortedVersions.slice(0, 5);
    displayVersions.forEach((version: string, index: number) => {
      const isLatest = index === 0;
      const marker = isLatest ? chalk.green(' (latest)') : '';
      console.log(`  • ${chalk.cyan(version)}${marker}`);
    });
    if (!options.detailed && sortedVersions.length > 5) {
      console.log(`  ${chalk.dim(`... and ${sortedVersions.length - 5} more (use --detailed to see all)`)}`);
    }
    console.log();
  }

  if (info.author) {
    console.log(chalk.bold('Author:'));
    console.log(`  ${info.author}`);
    console.log();
  }
}

function formatBytes(bytes: number): string {
  if (bytes === 0) return '0 Bytes';
  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return Math.round((bytes / Math.pow(k, i)) * 100) / 100 + ' ' + sizes[i];
}

File: go/cmd/prompd/explain.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"
)

func handleExplain() {
	if len(os.Args) < 3 {
		printExplainUsage()
		os.Exit(1)
	}

	target := os.Args[2]

	// Handle help flags
	if target == "-h" || target == "--help" {
		printExplainUsage()
		return
	}
	detailed := false
	sections := false
	registry := ""

	// Parse flags
	for i := 3; i < len(os.Args); i++ {
		switch os.Args[i] {
		case "-d", "--detailed":
			detailed = true
		case "-s", "--sections":
			sections = true
		case "-r", "--registry":
			if i+1 < len(os.Args) {
				registry = os.Args[i+1]
				i++
			}
		}
	}

	// Determine target type
	if strings.HasSuffix(target, ".prmd") {
		explainPrompdFile(target, detailed, sections)
	} else if strings.HasSuffix(target, ".pdpkg") {
		explainPackageFile(target, detailed)
	} else if strings.HasPrefix(target, "@") && strings.Contains(target, "/") {
		explainRegistryPackage(target, registry, detailed)
	} else {
		fmt.Println("ERROR: Invalid target. Must be a .prmd file, .pdpkg package, or registry package (@namespace/name)")
		os.Exit(1)
	}
}

func explainPrompdFile(filePath string, detailed bool, showSections bool) {
	// Check if file exists
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		fmt.Printf("ERROR: File not found: %s\n", filePath)
		os.Exit(1)
	}

	// Parse the file
	prompd, err := parsePrompdFile(filePath)
	if err != nil {
		fmt.Printf("ERROR: Failed to parse file: %v\n", err)
		os.Exit(1)
	}

	// Get file stats
	fileInfo, _ := os.Stat(filePath)

	fmt.Printf("\n📄 %s\n", filepath.Base(filePath))
	absPath, _ := filepath.Abs(filePath)
	fmt.Printf("   %s\n\n", absPath)

	// Metadata
	fmt.Println("Metadata:")
	fmt.Printf("  ID:          %s\n", prompd.Metadata.ID)
	if prompd.Metadata.Name != "" {
		fmt.Printf("  Name:        %s\n", prompd.Metadata.Name)
	}
	if prompd.Metadata.Version != "" {
		fmt.Printf("  Version:     %s\n", prompd.Metadata.Version)
	}
	if prompd.Metadata.Description != "" {
		fmt.Printf("  Description: %s\n", prompd.Metadata.Description)
	}
	fmt.Println()

	// Parameters
	if len(prompd.Metadata.Parameters) > 0 {
		fmt.Println("Parameters:")
		for _, param := range prompd.Metadata.Parameters {
			required := "optional"
			if param.Required {
				required = "required"
			}
			defaultVal := ""
			if param.Default != nil {
				defaultVal = fmt.Sprintf(" (default: %v)", param.Default)
			}
			fmt.Printf("  • %s (%s) %s%s\n", param.Name, param.Type, required, defaultVal)
			if param.Description != "" {
				fmt.Printf("    %s\n", param.Description)
			}
		}
		fmt.Println()
	}

	// Note: Section parsing not yet implemented in Go parser
	if showSections {
		fmt.Println("Note: Section display requires enhanced parser")
		fmt.Println()
	}

	// File info
	fmt.Println("File Info:")
	fmt.Printf("  Size:     %s\n", formatBytes(fileInfo.Size()))
	fmt.Printf("  Modified: %s\n", fileInfo.ModTime().Format(time.RFC1123))
	fmt.Println()
}

func explainPackageFile(filePath string, detailed bool) {
	// Check if file exists
	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		fmt.Printf("ERROR: Package not found: %s\n", filePath)
		os.Exit(1)
	}

	fileInfo, _ := os.Stat(filePath)

	fmt.Printf("\n📦 %s\n", filepath.Base(filePath))
	absPath, _ := filepath.Abs(filePath)
	fmt.Printf("   %s\n\n", absPath)

	fmt.Println("Package Info:")
	fmt.Printf("  Size:     %s\n", formatBytes(fileInfo.Size()))
	fmt.Printf("  Modified: %s\n", fileInfo.ModTime().Format(time.RFC1123))
	fmt.Println()

	if detailed {
		fmt.Println("Use 'prompd package validate' for detailed package inspection")
		fmt.Println()
	}
}

func explainRegistryPackage(packageName string, registry string, detailed bool) {
	fmt.Printf("Fetching package info for %s...\n\n", packageName)

	// Get package info from registry
	info, err := getPackageInfo(packageName, registry)
	if err != nil {
		fmt.Printf("ERROR: Failed to get package info: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("\n📦 %s\n\n", packageName)

	if info.Description != "" {
		fmt.Println(info.Description)
		fmt.Println()
	}

	fmt.Println("Package Info:")
	fmt.Printf("  Name:    %s\n", info.Name)
	if info.Version != "" {
		fmt.Printf("  Version: %s\n", info.Version)
	}
	fmt.Println()

	// Get versions
	versions, err := getPackageVersions(packageName, registry)
	if err == nil && len(versions) > 0 {
		fmt.Println("Available Versions:")
		displayCount := 5
		if detailed {
			displayCount = len(versions)
		}

		for i := 0; i < displayCount && i < len(versions); i++ {
			marker := ""
			if i == 0 {
				marker = " (latest)"
			}
			fmt.Printf("  • %s%s\n", versions[i], marker)
		}

		if !detailed && len(versions) > 5 {
			fmt.Printf("  ... and %d more (use --detailed to see all)\n", len(versions)-5)
		}
		fmt.Println()
	}

	if info.Author != "" {
		fmt.Println("Author:")
		fmt.Printf("  %s\n", info.Author)
		fmt.Println()
	}
}

func printExplainUsage() {
	fmt.Println(`Usage: prompd explain <target> [options]

Detailed information about files and packages

Arguments:
  <target>              .prmd file, .pdpkg package, or registry package (@namespace/name)

Options:
  -d, --detailed        Show detailed information
  -s, --sections        Show section content previews (for .prmd files)
  -r, --registry <name> Specify registry for package lookup

Examples:
  prompd explain example.prmd
  prompd explain example.prmd --sections
  prompd explain package.pdpkg
  prompd explain @namespace/package
  prompd explain @namespace/package --detailed`)
}

func formatBytes(bytes int64) string {
	const unit = 1024
	if bytes < unit {
		return fmt.Sprintf("%d B", bytes)
	}
	div, exp := int64(unit), 0
	for n := bytes / unit; n >= unit; n /= unit {
		div *= unit
		exp++
	}
	return fmt.Sprintf("%.1f %cB", float64(bytes)/float64(div), "KMGTPE"[exp])
}
